### DISCUSSION TITLE
CLI kills snakes in hazard before giving them a chance to eat

### DISCUSSION THREAD

**Discussion by exzizt** [BattlesnakeOfficial/feedback#77] [Bug Reports]
I think I've found a logic error in the rules code? Let me know what you think.

The royale ruleset is applied and then the standard set is applied, which is ok:
```
    if GameType == "royale" {
        _, err := royale.CreateNextBoardState(state, moves)
        if err != nil {
            log.Panic("[PANIC]: Error Producing Next Royale Board State")
            panic(err)
        }
    }
    state, err := ruleset.CreateNextBoardState(state, moves)
    if err != nil {
        log.Panic("[PANIC]: Error Producing Next Board State")
        panic(err)
    }
```
But the royale ruleset is improperly setting an "eliminated state" for snakes that starve in the hazard:
```
                if head == p {
                    // Snake is now out of bounds, reduce health
                    snake.Health = snake.Health - r.DamagePerTurn
                    if snake.Health < 0 {
                        snake.Health = 0
                    }
                    if r.StandardRuleset.snakeIsOutOfHealth(snake) {
                        snake.EliminatedCause = EliminatedByOutOfHealth
                    }
                }
```
This means that by the time the standard ruleset is applied, snakes can be "dead", even though they should be given the chance to eat food in the standard ruleset, and only should be considered "starved" once the "maybe eliminate snakes" function is called.
```
    err = r.maybeEliminateSnakes(nextState)
    if err != nil {
        return nil, err
    }
```

Fixing this logic error would also adhere to the intentions made by this comment about snakes being able to eat on their last turn:
```
    // bvanvugt: We specifically want this to happen before elimination for two reasons:
    // 1) We want snakes to be able to eat on their very last turn and still survive.
    // 2) So that head-to-head collisions on food still remove the food.
    //    This does create an artifact though, where head-to-head collisions
    //    of equal length actually show length + 1 and full health, as if both snakes ate.
    err = r.maybeFeedSnakes(nextState)
    if err != nil {
        return nil, err
    }
```

Killing the snakes that starve in the hazard sauce without giving them the chance to eat food breaks the rule :)

**Comment by bvanvugt** [comment]
@exzizt this is a wonderful catch, thank you for all the detail here. We've prioritized a fix for this internally, stay tuned.

**Comment by bvanvugt** [comment]
Should be fixed in [Rules v1.0.18](https://github.com/BattlesnakeOfficial/rules/releases/tag/v1.0.18)
